### PR TITLE
feat: add URL record format

### PR DESCRIPTION
**Pull request by gajus**
**What issue does this pull request resolve?**

Adds URL record format as defined in the [whatwg URL](https://url.spec.whatwg.org/) specification.

**What changes did you make?**

* I have added `URL` regex to `lib/compile/formats.js`.
* I have defined `url` property on `formats.fast` using `URL` regex.
* I have defined `url` property on `formats.full` using `URL` regex.

**Is there anything that requires more attention while reviewing?**

I am not familiar with the code base. If adding a format requires to add tests, please direct me to the relevant test files (or an example of a format test).

For one, I don't recommend adding tests for regex coverage, as it would require to cover quite a few cases (and thats been done already by Mathias, https://mathiasbynens.be/demo/url-regex).

**Whats the difference from the existing `uri` format?**

URL is a subset of URI specification.

Using the current format, `/test` is a valid URI, but not a valid URL.

### CODE REVIEW

**Comment by gajus** [comment]
Should this be not merged, it can be added by a user using:

```js
// For the source: https://gist.github.com/dperini/729294
// For test cases: https://mathiasbynens.be/demo/url-regex
const URL_RULE = /^(?:(?:https?|ftp):\/\/)(?:\S+(?::\S*)?@)?(?:(?!(?:10|127)(?:\.\d{1,3}){3})(?!(?:169\.254|192\.168)(?:\.\d{1,3}){2})(?!172\.(?:1[6-9]|2\d|3[0-1])(?:\.\d{1,3}){2})(?:[1-9]\d?|1\d\d|2[01]\d|22[0-3])(?:\.(?:1?\d{1,2}|2[0-4]\d|25[0-5])){2}(?:\.(?:[1-9]\d?|1\d\d|2[0-4]\d|25[0-4]))|(?:(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)(?:\.(?:[a-z\u00a1-\uffff0-9]-*)*[a-z\u00a1-\uffff0-9]+)*(?:\.(?:[a-z\u00a1-\uffff]{2,}))\.?)(?::\d{2,5})?(?:[/?#]\S*)?$/i;

ajv.addFormat('url', URL_RULE);
```

**Comment by epoberezkin** [comment]
@gajus it can be merged.
Please add the tests here: https://github.com/epoberezkin/ajv/blob/master/spec/tests/rules/format.json#L74
And docs: https://github.com/epoberezkin/ajv#formats (ideally with the link to json-schema spec  proposal :)

**Comment by gajus** [comment]
https://github.com/json-schema-org/json-schema-spec/issues/233

**Comment by gajus** [comment]
I have to disable the precommit hook to make the 625bed1 commit, because the tests are failing.

However, the rule is valid and the test cases are valid. https://jsbin.com/cepeqab/1/edit?js,output

How is the regex test executed on Ajv?

**Comment by epoberezkin** [comment]
your jsbin is somehow different - it fails for me: https://jsbin.com/duqoniqiha/1/edit?js,console

this fails: "http://a.b--c.de/"

**Comment by gajus** [comment]
Sorry, I forgot to remove 3 test cases that are not covered by the regex. That doesn't change the original problem.

**Comment by epoberezkin** [comment]
What's the original problem? I think these tests should pass so it's regex that is the problem, no?

File: lib/compile/formats.js [hunk 1]
```diff
@@ -9,6 +9,8 @@ var HOSTNAME = /^[0-9a-z](?:(?:[-0-9a-z]{0,61})?[0-9a-z])?(\.[0-9a-z](?:(?:[-0-9
...
 // For the source: https://gist.github.com/dperini/729294
 // For test cases: https://mathiasbynens.be/demo/url-regex
+// @todo Delete current URL in favour of the commented out URL rule when this issue is fixed https://github.com/eslint/eslint/issues/7983.
```
Comment (gajus) [inline lib/compile/formats.js]: I have learned that I am wrong:

1. It is not a bug in ESLint. It is a configuration issue. https://github.com/eslint/eslint/issues/7983#issuecomment-275257540
2. This cannot be used in the code until general ES6 adoption.

**Comment by gajus** [comment]
Is there anything else that needs to be done for this to get merged? (or would you rather close it, given that it is unlikely to be added to json-schema standard).

**Comment by objectiveSee** [comment]
I would like to see this merged in :) Commenting so I get notifications 💯

**Comment by gajus** [comment]
> I would like to see this merged in :) Commenting so I get notifications 💯

There is a subscribe button on the issue and 👍 to express your vote. Don't spam others on the issue.

**Pull request merged**